### ISSUE TITLE
Verify ERC20 contract address

### ISSUE THREAD

**Issue by erkarl** [ExchangeUnion/xud#897]
@offerm brought up a good point that we don't verify the contract address:
"if A has WETH/BTC and B has WETH/BTC it can be that they don't have the same WETH"

@kilrau:
I would want to verify the token address in two places:
- receiving order from peer, check that same ticker symbol has same token address than ours before adding to order book. This should catch the vast majority and prevents swaps to start and then fail. Potentially needs alert event and definitely log entry.
- right before swap, just in case of an edge case where an order with different token address made into the order book.

**Comment by erkarl** [comment]
My initial thoughts on this one are that we should handle the trading pair internally as <contract_address>/BTC or <contract_address>/<contract_address> for ERC20/ER20. Thoughts?

**Comment by kilrau** [comment]
And WETH would be just an alias which we use to represent given token address on the CLI/gRPC. Sounds good, prevents all 'cheating' and is essentially what eth does.

**Comment by sangaman** [comment]
> My initial thoughts on this one are that we should handle the trading pair internally as <contract_address>/BTC or <contract_address>/<contract_address> for ERC20/ERC20. Thoughts?

It makes some sense I think but I'm not sure it's necessary, as long we're never using the same currency symbol twice. In terms of the original issue, sanity swaps (#822) will make sure we're using the same token address for the advertised currencies. If we're not, the pair will remain dormant for that peer.

**Comment by kilrau** [comment]
> as long we're never using the same currency symbol twice

Wouldn't this require us to hardcode all ERC20 ticker symbol <-> contract addresses and hope that users don't change to cheat on these? Also wouldn't that eliminate the option to manually add ERC20 tokens (I picture two users manually adding the ticker symbol XUC, but with different contract addresses - how would the protocol handle this?). In the end ERC20 tokens are uniquely identified by their contract addresses. On raiden and full node level and that's what we should do too.

**Comment by sangaman** [comment]
I meant more that a particular node isn't using the same currency symbol twice, which is actually impossible currently. If we want to support the ability to trade multiple contract addresses for the same token/currency then that's a different story, but as far as I know that's not something we'd want to do (it'd certainly be horribly inefficient from a network perspective).

As is pending the sanity swaps PR, if two nodes tried to use different token addresses for the same currency, they'd fail on the sanity swap and they'd never exchange orders for that pair. That seems fine to me. Practically speaking, network participants need to agree on contract addresses to use for all the tokens, otherwise they won't be able to trade regardless of what we do within xud.

**Comment by kilrau** [comment]
> I meant more that a particular node isn't using the same currency symbol twice, which is actually impossible currently.

Gotcha.

> If we want to support the ability to trade multiple contract addresses for the same token/currency then that's a different story, but as far as I know that's not something we'd want to do (it'd certainly be horribly inefficient from a network perspective).

Indeed that would be wrong. Different contract address = different token.

> Practically speaking, network participants need to agree on contract addresses to use for all the tokens, otherwise they won't be able to trade regardless of what we do within xud.

But isn't that precisely the justification why we should only care about contract addresses for ERC20s as unique id? No agreement between nodes needed. A user can call the ticker symbol differently on her node, but the network only needs to care about the contract address. Sry, but I'm just really convinced that this is how it HAS to be.

To back this up, I just gave that a quick test with an ERC20 wallet:
* I added a new token with contract address `0xc324a2f6b05880503444451b8b27e6f9e63287cb`, name `XUC`, Symbol `XUC`, 18 decimals
* I added a new token once more with the same contract address `0xc324a2f6b05880503444451b8b27e6f9e63287cb`, name `ABC`, Symbol `ABC`, 12 decimals.

It should have definitely asked me, but the wallet simply replaced XUC with the new ABC token.

* Then I added a new token once more with a *DIFFERENT* contract address `0xc324a2f6b05880503444451b8b27e6f9e63287cb`, same name `ABC`, same Symbol `ABC`, same 12 decimals.

It added a second `ABC` token without replacing.

The contract address is the only id for a token and that's how `xud` should behave too.
